### PR TITLE
Modified StorageBearerTokenChallengeAuthorizationPolicy to use scope r…

### PR DESCRIPTION
**Pull request by seanmcc-msft**
…eturned by Bearer Challenge

### CODE REVIEW

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs [hunk 1]
```diff
@@ -81,6 +81,14 @@ private async ValueTask<bool> AuthorizeRequestOnChallengeInternalAsync(HttpMessa
                 // tenantId should be the guid as seen in this example: https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/authorize
                 tenantId = new Uri(authUri).Segments[1].Trim('/');
 
+                if (message.Response.Headers.TryGetValue(HttpHeader.Names.WwwAuthenticate, out string wwwAuthenticate)
+                    && wwwAuthenticate.Contains(Constants.ResourceId))
+                {
+                    string scope = wwwAuthenticate.Split(new string[] { Constants.ResourceId + "=" }, StringSplitOptions.None)[1];
```
Comment (christothes) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:87]: In this method, you already know the wwwAuth header is there. I'd also recommend that you use `AuthorizationChallengeParser.GetChallengeParameterFromResponse` to parse, as it is super efficient.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs]: fixed.

**Comment by seanmcc-msft** [comment]
TODO update common changelog.

File: sdk/storage/Azure.Storage.Common/tests/StorageBearerTokenChallengeAuthorizationPolicyTests.cs [hunk 2]
```diff
@@ -100,5 +102,46 @@ public async Task SendsUnUnAuthroizedRequestWhenEnableTenantDiscoveryIsTrue([Val
...
             Assert.True(transport.Requests.Skip(enableTenantDiscovery ? 1 : 0).All(r => r.Headers.TryGetValue("Authorization", out _)));
         }
+
+        [Test]
+        public async Task UsesScopeFromBearerChallange()
+        {
+            // Arrange
+            bool firstRequest = true;
+            string unauthorizedScope = "https://disk.compute.azure.com/.default";
+            string authorizedScope = "https://storage.azure.com";
```
Comment (christothes) [inline sdk/storage/Azure.Storage.Common/tests/StorageBearerTokenChallengeAuthorizationPolicyTests.cs:112]: For clarity I would recommend naming these something like:
```suggestion
            string initialMismatchedScope = "https://disk.compute.azure.com/.default";
            string serviceChallengeResponseScope = "https://storage.azure.com";
```

**Review by christothes — COMMENTED** [review]
Is there a live test that can be added to exercise this new scenario?

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs [hunk 3]
```diff
@@ -81,6 +81,10 @@ private async ValueTask<bool> AuthorizeRequestOnChallengeInternalAsync(HttpMessa
                 // tenantId should be the guid as seen in this example: https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/authorize
                 tenantId = new Uri(authUri).Segments[1].Trim('/');
 
+                string scope = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "resource_id");
+                scope += Constants.DefaultScope;
+                _scopes[0] = scope;
```
Comment (ryanerdmann) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:86]: This looks a little questionable to me.  The policy is expected to be thread-safe and shared across requests, but this is mutating instance state on a per-request basis.   (The `_scopes` variable is also annotated `readonly`, and even though that does not extend to the values of the array, it violates the principle of least surprise to see the elements change.)

While the `resource_id` is currently constant across all requests, I'm concerned about taking that dependency here.  

(As an aside, why is the `tenantId`  variable not named `_tenantId`?)

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs [hunk 4]
```diff
@@ -81,6 +81,10 @@ private async ValueTask<bool> AuthorizeRequestOnChallengeInternalAsync(HttpMessa
                 // tenantId should be the guid as seen in this example: https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/authorize
                 tenantId = new Uri(authUri).Segments[1].Trim('/');
 
+                string scope = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "resource_id");
```
[the remark this answers is not in the document]
Comment (ryanerdmann) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:84]: And, if this is now always overridden, should `scope` still be a parameter to the `StorageBearerTokenChallengeAuthorizationPolicy` ctor?

[the remark this answers is not in the document]
Comment (christothes) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:84]: scope is still passed to the base class to be used as the scope for the request to GetToken. The scope is only replaced here in the event of a challenge, which would not occur in the main line case.

[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:84]: `scope` is still being set to `https://storage.azure.com/.default` by default.  Most SDK customers will be using the scope scope anyway, this saves them the extra 401, and avoids having to change the default path for OAuth and re-record all the OAuth tests.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs (hunk 3 above)
Comment (christothes) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:86]: See how ChallengeBasedAuthenticationPolicy does it in a threadsafe way here: 

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L20

**Comment by seanmcc-msft** [comment]
> Is there a live test that can be added to exercise this new scenario?

@christothes, this could be kind of tricky since we are defaulting to the storage scope.  The managed disk scope if a pretty big edge case, and would require us to refactor our live test ARM template a bit.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:86]: In running local live tests, I've noticed the current behavior is that the client will continue sending the invalid scope and getting a bearer challenge for each request it sends, and then sending a follow up request with the correct scope.

Could I just switch scope to a local variable, as @ryanerdmann suggested?

Comment (christothes) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:86]: It would only be referenced locally when a challenge is received. For non-challenge scenarios, you'll need to supply the new scope somehow. ChallengeBasedAuthenticationPolicy handles this in its override of `AuthorizeRequestAsync` which calls `AuthorizeRequestInternal`

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:89]: Covered null case.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs]: I have updated the PR to pass _scopes to the base.  We may still have a thread safety issue though.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs [hunk 5]
```diff
@@ -49,21 +49,26 @@ public StorageBearerTokenChallengeAuthorizationPolicy(TokenCredential credential
...
         /// <inheritdoc />
         protected override ValueTask AuthorizeRequestAsync(HttpMessage message)
+            => AuthorizeRequestInternal(message, true);
+
+        private async ValueTask AuthorizeRequestInternal(HttpMessage message, bool async)
         {
             if (tenantId != null || !_enableTenantDiscovery)
             {
-                return base.AuthorizeRequestAsync(message);
+                TokenRequestContext context = new TokenRequestContext(_scopes, message.Request.ClientRequestId);
```
Comment (christothes) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs:62]: We should send the tenantId along as well here. 

```suggestion
                TokenRequestContext context = new TokenRequestContext(_scopes, message.Request.ClientRequestId, tenantId: tenantId);
```

File: sdk/storage/Azure.Storage.Common/tests/StorageBearerTokenChallengeAuthorizationPolicyTests.cs [hunk 6]
```diff
@@ -100,5 +102,46 @@ public async Task SendsUnUnAuthroizedRequestWhenEnableTenantDiscoveryIsTrue([Val
...
+            string initialMismatchedScope = "https://disk.compute.azure.com/.default";
+            string serviceChallengeResponseScope = "https://storage.azure.com";
+
+            string[] initialMismatchedScopes = new string[] { initialMismatchedScope };
+            string[] serviceChallengeResponseScopes = new string[] { serviceChallengeResponseScope + "/.default" };
+            MockCredential mockCredential = new MockCredential()
+            {
+                GetTokenCallback = (trc, _) =>
+                {
+                    Assert.AreEqual(serviceChallengeResponseScopes, trc.Scopes);
```
Comment (christothes) [inline sdk/storage/Azure.Storage.Common/tests/StorageBearerTokenChallengeAuthorizationPolicyTests.cs:120]: We should validate that the callback is only called once even if there are multiple requests sent to validate that we are caching properly.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/tests/StorageBearerTokenChallengeAuthorizationPolicyTests.cs:122]: fixed.

**Comment by christothes** [comment]
.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs [hunk 7]
```diff
@@ -49,21 +49,26 @@ public StorageBearerTokenChallengeAuthorizationPolicy(TokenCredential credential
...
+
+        private async ValueTask AuthorizeRequestInternal(HttpMessage message, bool async)
         {
             if (tenantId != null || !_enableTenantDiscovery)
             {
-                return base.AuthorizeRequestAsync(message);
+                TokenRequestContext context = new TokenRequestContext(_scopes, message.Request.ClientRequestId, tenantId: tenantId);
+                if (async)
+                {
+                    await base.AuthenticateAndAuthorizeRequestAsync(message, context).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs]: Fixed.

File: sdk/storage/Azure.Storage.Common/tests/StorageBearerTokenChallengeAuthorizationPolicyTests.cs [hunk 8]
```diff
@@ -100,5 +102,49 @@ public async Task SendsUnUnAuthroizedRequestWhenEnableTenantDiscoveryIsTrue([Val
...
+                    {
+                        true => challengeResponse,
+                        false => new MockResponse(200)
+                    };
+                    firstRequest = false;
+                    return response;
+                });
+
+            // Act
+            await SendGetRequest(transport, tokenChallengeAuthorizationPolicy, uri: new Uri("https://example.com"));
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/tests/StorageBearerTokenChallengeAuthorizationPolicyTests.cs:147]: Fixed.

**Review by christothes — APPROVED** [review]
